adapters/repos/db/migrator_shard_status_ops.go

```diff
@@ -0,0 +1,217 @@
+					err = fmt.Errorf("attempt to delete local fs for shard %s: %w", name, err)
+					ec.Add(err)
+					return err
+				}
+				return nil
+			}
+
+			idx.shardCreateLocks.Lock(name)
+			defer idx.shardCreateLocks.Unlock(name)
+
+			idx.shards.LoadAndDelete(name)
+
+			if err := shard.drop(); err != nil {
+				ec.Add(err)
+			}
+			return nil
+		})
+	}
+	eg.Wait()
+}
+
+func (m *Migrator) freeze(ctx context.Context, idx *Index, class string, freeze []string, ec *errorcompounder.ErrorCompounder) {
+	if m.cloud == nil {
+		ec.Add(fmt.Errorf("offload to cloud module is not enabled"))
+		return
+	}
+
+	if m.cluster == nil {
+		ec.Add(fmt.Errorf("no cluster exists in the migrator"))
+		return
+	}
+
+	idx.backupMutex.RLock()
+	defer idx.backupMutex.RUnlock()
+
+	eg := enterrors.NewErrorGroupWrapper(m.logger)
+	eg.SetLimit(_NUMCPU * 2)
+
+	for _, name := range freeze {
+		name := name
+		originalStatus := models.TenantActivityStatusHOT
+		eg.Go(func() error {
+			shard := func() ShardLike {
+				idx.shardInUseLocks.Lock(name)
+				defer idx.shardInUseLocks.Unlock(name)
+
+				return idx.shards.Load(name)
+			}()
+
+			if shard == nil {
+				// shard already does not exist or inactive
+				originalStatus = models.TenantActivityStatusCOLD
+			}
+
+			idx.shardCreateLocks.Lock(name)
+			defer idx.shardCreateLocks.Unlock(name)
+
+			if shard != nil {
+				if err := shard.HaltForTransfer(ctx); err != nil {
+					ec.Add(err)
+					return fmt.Errorf("attempt to mark begin offloading: %w", err)
+				}
+			}
+
+			enterrors.GoWrapper(func() {
+				cmd := command.TenantProcessRequest{
+					Node:   m.nodeId,
+					Action: command.TenantProcessRequest_ACTION_FREEZING,
+				}
+				err := m.cloud.Upload(ctx, class, name, m.nodeId)
+				if err != nil {
+					ec.Add(fmt.Errorf("uploading error: %w", err))
+					cmd.Process = &command.TenantsProcess{
+						Tenant: &command.Tenant{
+							Name:   name,
+							Status: originalStatus,
+						},
+						Op: command.TenantsProcess_OP_ABORT,
+					}
+				} else {
+					cmd.Process = &command.TenantsProcess{
+						Tenant: &command.Tenant{
+							Name:   name,
+							Status: models.TenantActivityStatusFROZEN,
+						},
+						Op: command.TenantsProcess_OP_DONE,
+					}
+				}
+
+				if _, err := m.cluster.UpdateTenantsProcess(class, &cmd); err != nil {
+					ec.Add(fmt.Errorf("UpdateTenantsProcess error: %w", err))
+				}
+			}, idx.logger)
+
+			return nil
+		})
+	}
+	eg.Wait()
+}
+
+func (m *Migrator) unfreeze(ctx context.Context, idx *Index, class string, unfreeze []string, ec *errorcompounder.ErrorCompounder) {
+	if m.cloud == nil {
+		ec.Add(fmt.Errorf("offload to cloud module is not enabled"))
+		return
+	}
+
+	if m.cluster == nil {
+		ec.Add(fmt.Errorf("no cluster exists in the migrator"))
+		return
+	}
+
+	idx.backupMutex.RLock()
+	defer idx.backupMutex.RUnlock()
+
+	eg := enterrors.NewErrorGroupWrapper(m.logger)
+	eg.SetLimit(_NUMCPU * 2)
+
+	for _, name := range unfreeze {
+		split := strings.Split(name, "#")
```

Comment:
Where is the `#` delineator coming from?

Reply:
yeah, good question. I've to add a comment here to describe it, this shall be coming from raft to identify which node path shall it gets it's data from. it's made because nodeIDs could be changed on download

Reply:
> `# is a delineator shall come from RAFT and it's away e.g. tenant1-node1`

Should this say `tenant1#node1`?

Reply:
yesss  😵‍💫

Reply:
thanks 😁